contrib/lib-autopilot.py

```diff
@@ -0,0 +1,289 @@
+* include more statistics of the network: 
+* allow autopilots of various nodes to exchange some information
+* exchange algorithms if the network grows.
+* include better handling for duplicates and existing channels
+* cap number of channels for well connected nodes. 
+"""
+
+import logging
+
+from lightning.lightning import LightningRpc
+import json
+import time
+import pickle
+
+import networkx as nx
+
+import heapq
+from operator import itemgetter
+import random
+import math
+
+from os.path import expanduser
+
+class Autopilot():
+    __rpc_interface = None
+    G = None 
+    __logger = None
+        
+    def __init__(self):
+        self.__add_logger()
+        #FIXME: find out where the config file is placed:
+        self.__rpc_interface = LightningRpc(expanduser("~")+"/.lightning/lightning-rpc")
+
+        try:
+            self.__logger.info("Try to load graph from file system at: data/networkx_graph")
+            with open("data/networkx_graph","rb") as infile:
+                self.G = pickle.load(infile)
+                self.__logger.info("Successfully restored the lightning network graph from data/networkx_graph")
+        except FileNotFoundError:
+            self.__logger.info("load the graph from the peers of the lightning network")
+        
+            self.G = nx.Graph()
+            if self.__load_nodes()==False:
+                self.__logger.info("can not download nodes from the network and initialize the networkx Graph")
+            if self.__load_edges()==False:
+                self.__logger.info("cann not download the channels from the network and initialize the networkx Graph")
+            else:
+                with open("data/networkx_graph", "wb") as outfile:
+                    pickle.dump(self.G,outfile,pickle.HIGHEST_PROTOCOL)
+            
+    def __add_logger(self):
+        """ initiates the logging service for this class """
+        #FIXME: adapt to the settings that are proper for you
+        self.__logger = logging.getLogger('lib-autopilot')
+        self.__logger.setLevel(logging.INFO)
+        ch = logging.StreamHandler()
+        ch.setLevel(logging.INFO)
+        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+        ch.setFormatter(formatter)
+        self.__logger.addHandler(ch)  
+        # TODO: CHANGE from Console to file handler
+        # fh = logging.FileHandler('lib-autopilot.log')
+        # fh.setLevel(logging.DEBUG)
+        #fh.setFormatter(formatter)
+        #self.__logger.addHandler(fh)
+        
+    def __load_nodes(self):
+        #FIXME: it is a real problem that we don't know how many nodes there could be
+        nodes = []
+        try:
+            self.__logger.info("Attempt RPC-call to download nodes from the lightning network")
+            while len(nodes) == 0: 
+                peers = self.__rpc_interface.listpeers()["peers"]
+                if len(peers) < 1:
+                    #FIXME: do we have seed nodes?
```

Comment:
Resolving `lseed.bitcoinstats.com` gives you 20 nodes listening on the default port, no guarantee they'll actually be reachable atm though (haven't implemented liveness check yet).

Reply:
how do I get the nodeids? when running a dns lookup I only get ip addresses. Renes-iMac:Bitcoin rpickhardt$ nslookup lseed.bitcoinstats.com
Server:		192.168.178.1
Address:	192.168.178.1#53

Non-authoritative answer:
Name:	lseed.bitcoinstats.com
Address: 100.38.11.146
Name:	lseed.bitcoinstats.com
Address: 70.29.71.171
Name:	lseed.bitcoinstats.com
Address: 88.152.223.194
Name:	lseed.bitcoinstats.com
Address: 5.9.2.145
Name:	lseed.bitcoinstats.com
Address: 46.38.245.95
Name:	lseed.bitcoinstats.com
Address: 188.166.9.103
Name:	lseed.bitcoinstats.com
Address: 85.214.212.104
Name:	lseed.bitcoinstats.com
Address: 35.185.13.214
Name:	lseed.bitcoinstats.com
Address: 165.227.7.29
Name:	lseed.bitcoinstats.com
Address: 178.192.63.108
Name:	lseed.bitcoinstats.com
Address: 84.245.27.209
Name:	lseed.bitcoinstats.com
Address: 13.92.254.226
Name:	lseed.bitcoinstats.com
Address: 159.203.122.47
Name:	lseed.bitcoinstats.com
Address: 94.176.237.167
Name:	lseed.bitcoinstats.com
Address: 89.70.225.66
Name:	lseed.bitcoinstats.com
Address: 178.254.20.84
Name:	lseed.bitcoinstats.com
Address: 206.189.170.76
Name:	lseed.bitcoinstats.com
Address: 98.209.138.74
Name:	lseed.bitcoinstats.com
Address: 85.11.37.130
Name:	lseed.bitcoinstats.com
Address: 192.168.2.1
Name:	lseed.bitcoinstats.com
Address: 35.203.126.128
Name:	lseed.bitcoinstats.com
Address: 85.216.75.225

Reply:
You can get the nodeids by issuing an `SRV` request first, and then resolving those virtual names:

```
dig lseed.bitcoinstats.com SRV @139.59.143.87 +short
10 10 9735 ln1qfffmd5l6t4axyn0keh6lg35ls65g3m6y02snl5na53fhv8f9e8msyxzk4z.lseed.bitcoinstats.com.
10 10 9735 ln1qtgzgvmx02mr9gcd23lrj2k72ya766vg6pskjnachg86f4c5m0tmwq34lqk.lseed.bitcoinstats.com.
10 10 9735 ln1qw2rjmnw3nudn7gausnfmlfwssttjwu3etnp8yzhcu5rk337f0arzc2m2h8.lseed.bitcoinstats.com.
...
```

```
dig +short ln1qfffmd5l6t4axyn0keh6lg35ls65g3m6y02snl5na53fhv8f9e8msyxzk4z.lseed.bitcoinstats.com AAAA 
2a02:c207:2016:2394::1
2001:470:8:2e1::43
2001:b030:2422::208d
```

Reply:
The nodeid is encoded as bech32 in order to fit into a single domain segment.